airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/up_to_date/pipeline.py

```diff
@@ -0,0 +1,366 @@
+            )
+        if POETRY_LOCK_FILE not in connector_dir_entries or POETRY_TOML_FILE not in connector_dir_entries:
+            return StepResult(
+                step=self,
+                status=StepStatus.SKIPPED,
+                stderr="The connector requires poetry.",
+            )
+
+        if not self.context.connector.metadata or not self.context.connector.metadata.get("connectorBuildOptions", {}).get("baseImage"):
+            return StepResult(
+                step=self,
+                status=StepStatus.SKIPPED,
+                stderr="The connector can't be updated because it does not have a base image defined in the metadata.",
+            )
+
+        return StepResult(
+            step=self,
+            status=StepStatus.SUCCESS,
+        )
+
+
+class UpdatePoetry(Step):
+    context: ConnectorContext
+    dev: bool
+    specified_versions: dict[str, str]
+
+    title = "Update versions of libraries in poetry."
+
+    def __init__(self, context: PipelineContext, dev: bool, specific_dependencies: List[str]) -> None:
+        super().__init__(context)
+        self.dev = dev
+        self.specified_versions = parse_specific_dependencies(specific_dependencies)
+
+    async def _run(self) -> StepResult:
+        base_image_name = self.context.connector.metadata["connectorBuildOptions"]["baseImage"]
+        base_container = self.dagger_client.container(platform=LOCAL_BUILD_PLATFORM).from_(base_image_name)
+        connector_container = await with_python_connector_installed(
+            self.context,
+            base_container,
+            str(self.context.connector.code_directory),
+        )
+
+        try:
+            before_versions = await get_poetry_versions(connector_container)
+            before_main = await get_poetry_versions(connector_container, only="main")
+
+            if self.specified_versions:
+                for package, dep in self.specified_versions.items():
+                    self.logger.info(f"  Specified: poetry add {dep}")
+                    if package in before_main:
+                        connector_container = await connector_container.with_exec(["poetry", "add", dep])
+                    else:
+                        connector_container = await connector_container.with_exec(["poetry", "add", dep, "--group=dev"])
+            else:
+                current_cdk_version = before_versions.get("airbyte-cdk") or None
+                if current_cdk_version:
+                    # We want the CDK pinned exactly so it also works as expected in PyAirbyte and other `pip` scenarios
+                    new_cdk_version = pick_airbyte_cdk_version(current_cdk_version, self.context)
+                    self.logger.info(f"Updating airbyte-cdk from {current_cdk_version} to {new_cdk_version}")
+                    if new_cdk_version > current_cdk_version:
+                        connector_container = await connector_container.with_exec(["poetry", "add", f"airbyte-cdk=={new_cdk_version}"])
```

Comment:
Can we remove this `else` block and instantiate a second `UpdatePoetry` step in the `run_up_to_date_pipeline`, the pipeline would pick the right cdk version (according to inputs or by fetching the latest version).
I would remove "business" logic from the `Step` and make it more reusable - self contained.

In other words: can we make `self.specified_versions` not optional 😄

Reply:
Maybe. But I don't thin kthe eventual `cron` script will specify the versions for all the updates that should be `poetry update` --- `specified_versions` was really only for local work. "Let's change try changing the semantic version of pytest" kind of stuff.